src/legacy/core_plugins/console/np_ready/public/lib/ace_token_provider/token_provider.ts

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { IEditSession, TokenInfo } from 'brace';
+import { TokensProvider, Token, Position } from '../../interfaces';
+
+interface ActualTokenInfo extends TokenInfo {
+  type: string;
+}
+
+const toToken = (lineNumber: number, tokens: ActualTokenInfo[]): Token[] => {
+  let acc = '';
+  return tokens.map(token => {
+    const column = acc.length + 1;
+    acc += token.value || '';
+    return {
+      type: token.type,
+      value: token.value,
+      position: {
+        lineNumber,
+        column,
+      },
+    };
+  });
+};
+
+export class AceTokensProvider implements TokensProvider {
+  constructor(private readonly session: IEditSession) {}
+
+  getTokens(lineNumber: number): Token[] | null {
+    if (lineNumber < 1) return null;
+    const tokens: ActualTokenInfo[] = this.session.getTokens(lineNumber - 1) as any;
+    if (!tokens || !tokens.length) {
+      const lineCount = this.session.doc.getAllLines().length;
+      if (lineNumber > 0 && lineCount > lineNumber) {
```

Comment:
Could you explain this special case for me? Why are we returning `[]` instead of `null`?

`lineNumber > 0` seems redundant given the check on line 47. And I speak from a position of ignorance so I might be way off, but `lineCount > lineNumber` seems like the consumer is trying to access a line number that doesn't exist. If this is the case, then can't we assume that line doesn't contain tokens? In this case can we return on line 48 before getting the tokens?

Reply:
Regarding the unnecessary `lineNumber > 0` check, you're 100% correct. Will update!

So what this logic is trying to express is that if the lineNumber the user provided is a non-zero positive value (which we established on line 47) AND this lineNumber is within the document's line count (i.e., `lineCount > lineNumber`, which should actually be `lineCount >= lineNumber` since both are 1-indexed) but ace gave us `null` it means we are on an empty line. The TokenProvider reserves `null` as special value meaning you are outside of the document range. Seeing that in this special case, ace gave us no tokens but we are inside of the document line count range, we return an empty array to tell consumers of the token provider we are inside of the document, but have no tokens for this existent line.

I'll add this to the interface documentation!

Reply:
Thanks that makes sense to me! Personally, I feel like we could express this intention through the code by putting the bounds-checks first. WDYT?

```js
getTokens(lineNumber: number): Token[] | null {
  if (lineNumber < 1) return null;

  const lineCount = this.session.doc.getAllLines().length;
  if (lineNumber > lineCount) return null;

  const tokens: ActualTokenInfo[] = this.session.getTokens(lineNumber - 1) as any;
  if (!tokens || !tokens.length) {
    // We're on an empty line.
    return [];
  }
  
  return toToken(lineNumber, tokens);
}
```